ambry-api/src/main/java/com/github/ambry/quota/Chargeable.java

```diff
@@ -0,0 +1,39 @@
+/**
+ * Copyright 2021 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.quota;
+
+/**
+ * A {@link Chargeable} is an operation that can be charged against quota.
+ */
+public interface Chargeable {
+
+  /**
+   * Charge the cost of the operation against relevant quota if usage is within quota.
+   *
+   * @return {@code true} if usage is within quota and quota was charged. {@code false} otherwise.
+   */
+  boolean checkAndCharge();
```

Comment:
Thank you for describing in detail on how this interface would be used! Just a question. Since chunks can belong to different partition classes, is it possible that chunk belonging to a partition class with more local replicas get more bandwidth (compared to chunk belonging to partition class with lesser local replicas) since they would have more requests sent from frontend to server for same charge.

Reply:
Each chunk will be charged just once. So, while frontend can send multiple rest requests to servers, for each chunk charging will happen only once.
In the implementation, each chunk class (like PutChunk) will maintain a state called `AtomicBoolean isCharged`. First time the `charge` method is called, it will also set `isCharged` to true. When more calls are made for the `PutChunk`, it will check if `isCharged` is already true, and if yes, then it will not charge again.